libbeat/docs/security/api-keys.asciidoc

```diff
@@ -0,0 +1,116 @@
+[role="xpack"]
+[[beats-api-keys]]
+=== Grant access using API keys
+
+Instead of using usernames and passwords you can use API keys to grant
+access to {es} resources. API keys can be set to expire at a certain time
+and explicitly invalidated. Any user with the `manage_api_key` or `manage_own_api_key`
+cluster privilege can create API keys.
+
+See the {es} API key documentation for more information:
+
+* {ref}/security-api-create-api-key.html[Create API key]
+* {ref}/security-api-get-api-key.html[Get API key information]
+* {ref}/security-api-invalidate-api-key.html[Invalidate API key]
+
+Every {beatname_uc} instance is typically sending both collected data and monitoring
+information to {es}. If you are sending both to the same cluster, you can use the same
+API key. For different clusters, you will need an API key per cluster.
+
+NOTE: For security reasons, we recommend using a unique API key per Beat instance.
+You can create as many API keys per user as necessary.
+
+[[beats-api-key-publish]]
+==== Create an API key for publishing
+To create an API key to use for writing data to {es}, use the
+{ref}/security-api-create-api-key.html[Create API key API], for example:
+
+[source,console,subs="attributes,callouts"]
+------------------------------------------------------------
+POST /_security/api_key
+{
+  "name": "{beat_default_index_prefix}_host001", <1>
+  "role_descriptors": {
+    "{beat_default_index_prefix}_writer": { <2>
+      "cluster": ["monitor", "read_ilm"],
+      "index": [
+        {
+          "names": ["{beat_default_index_prefix}-*"],
+          "privileges": ["view_index_metadata", "create_doc"]
+        }
+      ]
+    }
+  }
+}
+------------------------------------------------------------
+<1> name of the API key
+<2> granted privileges, see <<feature-roles>>
+
+The return value will look something like this:
+
+[source,console-result,subs="attributes,callouts"]
+--------------------------------------------------
+{
+  "id":"TiNAGG4BaaMdaH1tRfuU", <1>
+  "name":"{beat_default_index_prefix}_host001",
+  "api_key":"KnR6yE41RrSowb0kQ0HWoA" <2>
+}
+--------------------------------------------------
+<1> unique id for this API key
+<2> generated API key
+
+You can now use this API key in your {beatname_uc}.yml configuration file like this:
+["source","yml",subs="attributes"]
+--------------------
+output.elasticsearch:
+  api_key: TiNAGG4BaaMdaH1tRfuU:KnR6yE41RrSowb0kQ0HWoA <1>
+--------------------
+<1> format is `id:api_key` (note the colon in the middle)
```

Comment:
```suggestion
<1> format is `id:api_key` (note the colon in `api_key`)
```

Reply:
There are two colons, so this statement is a bit ambiguous without more clarification.

Reply:
There are two colons?

Reply:
Sorry, my comment was confusing.

I meant that the line in the example contains two colons. The way my eye switches between a callout and example,  I apply everything I see in the callout to the example, including the parenthetical statement. It causes a weird disconnect for me. Might be subjective, but anything that causes the reader to stumble isn't ideal.

My suggested edit wasn't great, either, because we use  `api_key` as the setting name and in the format example. I think that's weird.

TBH, I don't think you need to point out the colon here because it's shown in the example. 
Instead, it would be useful to say: "The format is `id:key`, where `id` is the ID returned by the Create API key API, and `key` is the API key.

Reply:
Oy "returned by the Create API key API" is a mouthful. Maybe: "The format is `id:key`, where `id` is the ID returned when you created the API key, and `key` is the API key.

Reply:
Ok, I see. I used `id:api_key` because that's what the [Elasticsearch Create API key API](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-api-key.html) returns. There is no `key` return value, which I think could confuse users ("what is `key`, where do I get it from?"). We could call out explicitly that these are two fields from the ES API output. Wdyt?

Reply:
Yah that would probably help!

Reply:
Another thought.... The fundamental problem here is that we use `api_key` to mean two different things: just the key, or the key with the ID.  It might be too late to change, but it might help to use a different name for the setting. Maybe something like `api_key_pair`? Maybe we already have a precedent for that.

Reply:
I've changed the callouts to `Format is id:api_key (as returned by Create API key)`. 

> It might be too late to change, but it might help to use a different name for the setting. Maybe something like `api_key_pair`? Maybe we already have a precedent for that.

It's not too late :-). I agree it's not ideal. This problem is already inherent in Elasticsearch's API, where the API endpoint is `/_security/api_key`, but then an API key has both an `id` and an actual `api_key`. Personally, I would not want to introduce another configuration parameter with a name that does not appear anywhere else.

Reply:
That's perfect! 

You're right; the API does have the same problem, so let's leave the config name as-is.